### PR TITLE
perf (FunLike.Basic): beta reduce CoeFun.coe 

### PR DESCRIPTION
**Pull request by mattrobball**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

File: Mathlib/Algebra/Algebra/RestrictScalars.lean [hunk 1]
```diff
@@ -216,9 +216,8 @@ theorem RestrictScalars.ringEquiv_map_smul (r : R) (x : RestrictScalars R S A) :
 /-- `R ⟶ S` induces `S-Alg ⥤ R-Alg` -/
 instance RestrictScalars.algebra : Algebra R (RestrictScalars R S A) :=
   { (algebraMap S A).comp (algebraMap R S) with
-    smul := (· • ·)
-    commutes' := fun _ _ ↦ Algebra.commutes' _ _
-    smul_def' := fun _ _ ↦ Algebra.smul_def' _ _ }
+    commutes' := fun _ _ ↦ Algebra.commutes' (A := A) _ _
+    smul_def' := fun _ _ ↦ Algebra.smul_def' (A := A) _ _ }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Algebra/RestrictScalars.lean:220]: Why was this change needed?

Comment (mattrobball) [reply Mathlib/Algebra/Algebra/RestrictScalars.lean:220]: L219: I like deleting these. It is independent of anything else. 
L220-221: I've seen elaboration struggle when the terms get less transparent/more compact. Both here and in other reductions of terms.

File: Mathlib/Algebra/Lie/DirectSum.lean [hunk 2]
```diff
@@ -197,7 +197,6 @@ theorem lie_of_of_eq [DecidableEq ι] {i j : ι} (hij : j = i) (x : L i) (y : L
     lieAlgebraOf_apply]
 #align direct_sum.lie_of_of_eq DirectSum.lie_of_of_eq
 
-@[simp]
 theorem lie_of [DecidableEq ι] {i j : ι} (x : L i) (y : L j) :
     ⁅of L i x, of L j y⁆ = if hij : j = i then lieAlgebraOf R ι L i ⁅x, hij.recOn y⁆ else 0 := by
   by_cases hij : j = i
```
Comment (eric-wieser) [inline Mathlib/Algebra/Lie/DirectSum.lean:202]: #8388
